### PR TITLE
Migrate `print-target-list` to `rmake` and  `print-calling-convention` to ui-test

### PR DESCRIPTION
**Pull request by rejyr**
Part of #121876.

r? @jieyouxu

try-job: x86_64-gnu-llvm-18
try-job: test-various
try-job: armhf-gnu
try-job: aarch64-apple
try-job: i686-mingw
try-job: x86_64-msvc

### CODE REVIEW

**Comment by tgross35** [comment]
Could you add checks for trailing newlines in the output? Just since we have missed them before (https://github.com/rust-lang/rust/pull/127512/files)

File: tests/run-make/print-target-list/rmake.rs
Comment (Rejyr) [inline tests/run-make/print-target-list/rmake.rs:10]: > The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9982698975/job/27588945618) [(plain)](https://github.com/rust-lang/rust/commit/65cab6fbaa4967f34a9b3584c2b97132ecdc712d/checks/27588945618/logs)
> Click to see the possible cause of the failure (guessed by this bot)
> 
> ```
> # snip
> error: could not create LLVM TargetMachine for triple: csky-unknown-linux-gnuabiv2: No available targets are compatible with triple "csky-unknown-linux-gnuabiv2"
> # snip
> ```

Apparently the target `csky-unknown-linux-gnuabiv2` is not found?

Comment (jieyouxu) [reply tests/run-make/print-target-list/rmake.rs:10]: I opened an issue https://github.com/rust-lang/rust/issues/127895 to ask its target maintainer (and other people) for more information regarding expected behavior here.

**Review by jieyouxu — CHANGES_REQUESTED** [review]

**Comment by jieyouxu** [comment]
Marking this as blocked on the pending discussions in #127895.
@rustbot blocked (waiting for more info from #127895)

File: tests/run-make/print-calling-conventions/rmake.rs
Comment (jieyouxu) [reply tests/run-make/print-calling-conventions/rmake.rs:1]: Suggestion: also can this become a UI test? May need some experimenting as to why it can't.

Comment (Rejyr) [reply tests/run-make/print-calling-conventions/rmake.rs:1]: Yes, this works as a UI test.

File: tests/run-make/print-target-list/rmake.rs
Comment (jieyouxu) [reply tests/run-make/print-target-list/rmake.rs:10]: Hm, this test might need a new directive `//@ needs-all-llvm-components` that checks if `COMPILETEST_REQUIRE_ALL_LLVM_COMPONENTS` is set. Let me get back to you in a bit.

https://github.com/rust-lang/rust/blob/b01a977b0737e177f9b03949b908c250020d0119/src/tools/compiletest/src/header.rs#L1412-L1430

File: tests/run-make/print-target-list/rmake.rs
Comment (jieyouxu) [inline tests/run-make/print-target-list/rmake.rs:3]: Suggestion: for now, let's try specifying that we need the llvm components:

```
//@ needs-llvm-components: aarch64 arm avr bpf csky hexagon i386 i586 i686
loongarch64 m68k mips msp430 nvptx64 powerpc riscv s390x sparc thumb wasm x86 xtensa
```

File: src/tools/tidy/src/target_specific_tests.rs
Comment (Rejyr) [inline src/tools/tidy/src/target_specific_tests.rs:20]: Previously unknown LLVM components.

File: src/tools/tidy/src/target_specific_tests.rs
Comment (Rejyr) [inline src/tools/tidy/src/target_specific_tests.rs:89]: Specifying `needs-llvm-components` without a target fails tidy.

**Review by rejyr — COMMENTED** [review]
Testing adding `needs-llvm-components` to `run-make/print-target-list`.

**Comment by jieyouxu** [comment]
Let me sit on this for a bit to think of a better way to do this...
@rustbot review

File: tests/run-make/print-target-list/Makefile
Comment (jieyouxu) [inline tests/run-make/print-target-list/Makefile:8]: Problem: I found out why we're struggling with this test. It's because this Makefile isn't actually checking what it says it's checking...

```make
# target specifications
all:
	for target in $(shell $(BARE_RUSTC) --print target-list); do \
		$(BARE_RUSTC) --target $$target --print sysroot || exit 1; \
	done
```

if we force an exit, we'll see the makefile actually fails in CI with the exact same error (https://github.com/rust-lang/rust/pull/128073#issuecomment-2266381093).

File: tests/run-make/print-target-list/rmake.rs
Comment (jieyouxu) [inline tests/run-make/print-target-list/rmake.rs:5]: Suggestion: not needed since I found out why Makefile wasn't failing

File: tests/run-make/print-target-list/rmake.rs
Comment (jieyouxu) [inline tests/run-make/print-target-list/rmake.rs:13]: Suggestion: we should skip the check (and leave a FIXME here) if the target is experimental, i.e.

https://github.com/rust-lang/rust/blob/1df0458781d6fd753a68c4cdc4de5313b1635dbd/src/bootstrap/src/core/build_steps/llvm.rs#L325-L328

skip if one of AVR, M68k, CSKY, Xtensa.

**Review by jieyouxu — CHANGES_REQUESTED** [review]
I finally found out why we were struggling...

File: tests/run-make/print-target-list/rmake.rs
Comment (jieyouxu) [reply tests/run-make/print-target-list/rmake.rs:5]: Suggestion: we should revert the tidy changes since they're no longer necessary.

**Review by jieyouxu — APPROVED** [review]
Cool, I'll run a  couple of try job to double-check.

**Pull request merged**
